File: java/client/src/main/java/org/apache/rocketmq/client/java/metrics/MessageMeterInterceptor.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.java.metrics;

import com.google.common.base.Stopwatch;
import io.opentelemetry.api.common.Attributes;
import java.time.Duration;
import java.util.List;
import java.util.Optional;
import org.apache.rocketmq.client.apis.consumer.PushConsumer;
import org.apache.rocketmq.client.apis.consumer.SimpleConsumer;
import org.apache.rocketmq.client.java.hook.Attribute;
import org.apache.rocketmq.client.java.hook.AttributeKey;
import org.apache.rocketmq.client.java.hook.MessageHookPoints;
import org.apache.rocketmq.client.java.hook.MessageHookPointsStatus;
import org.apache.rocketmq.client.java.hook.MessageInterceptor;
import org.apache.rocketmq.client.java.hook.MessageInterceptorContext;
import org.apache.rocketmq.client.java.impl.Client;
import org.apache.rocketmq.client.java.message.GeneralMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class MessageMeterInterceptor implements MessageInterceptor {
    static final AttributeKey<Stopwatch> SEND_STOPWATCH_KEY = AttributeKey.create("send_stopwatch");
    static final AttributeKey<Stopwatch> CONSUME_STOPWATCH_KEY = AttributeKey.create("consume_stopwatch");

    private static final Logger log = LoggerFactory.getLogger(MessageMeterInterceptor.class);

    private final Client client;
    private final ClientMeterManager meterManager;

    public MessageMeterInterceptor(Client client, ClientMeterManager meterManager) {
        this.client = client;
        this.meterManager = meterManager;
    }

    private void doBeforeSendMessage(MessageInterceptorContext context) {
        // Record the time before sending message.
        context.putAttribute(SEND_STOPWATCH_KEY, Attribute.create(Stopwatch.createStarted()));
    }

    private void doAfterSendMessage(MessageInterceptorContext context, List<GeneralMessage> messages) {
        final Attribute<Stopwatch> stopwatchAttr = context.getAttribute(SEND_STOPWATCH_KEY);
        if (null == stopwatchAttr) {
            // Should never reach here.
            return;
        }
        for (GeneralMessage message : messages) {
            final Duration duration = stopwatchAttr.get().elapsed();
            InvocationStatus status = MessageHookPointsStatus.OK.equals(context.getStatus()) ?
                InvocationStatus.SUCCESS : InvocationStatus.FAILURE;
            Attributes attributes = Attributes.builder().put(MetricLabels.TOPIC, message.getTopic())
                .put(MetricLabels.CLIENT_ID, client.getClientId().toString())
                .put(MetricLabels.INVOCATION_STATUS, status.getName()).build();
            meterManager.record(HistogramEnum.SEND_COST_TIME, attributes, duration.toMillis());
        }
    }

    private void doAfterReceiveMessage(List<GeneralMessage> messages) {
        if (messages.isEmpty()) {
            return;
        }
        String consumerGroup = null;
        if (client instanceof PushConsumer) {
            consumerGroup = ((PushConsumer) client).getConsumerGroup();
        }
        if (client instanceof SimpleConsumer) {
            consumerGroup = ((SimpleConsumer) client).getConsumerGroup();
        }
        if (null == consumerGroup) {
            log.error("[Bug] consumerGroup is not recognized, clientId={}", client.getClientId());
            return;
        }
        final GeneralMessage message = messages.iterator().next();
        final Optional<Long> optionalTransportDeliveryTimestamp = message.getTransportDeliveryTimestamp();
        if (!optionalTransportDeliveryTimestamp.isPresent()) {
            return;
        }
        final long transportDeliveryTimestamp = optionalTransportDeliveryTimestamp.get();
        final long currentTimeMillis = System.currentTimeMillis();
        final long latency = currentTimeMillis - transportDeliveryTimestamp;
        if (0 > latency) {
            log.debug("latency is negative, latency={}ms, currentTimeMillis={}, transportDeliveryTimestamp={}",
                latency, currentTimeMillis, transportDeliveryTimestamp);
            return;
        }
        final Attributes attributes = Attributes.builder().put(MetricLabels.TOPIC, message.getTopic())
            .put(MetricLabels.CONSUMER_GROUP, consumerGroup)
            .put(MetricLabels.CLIENT_ID, client.getClientId().toString()).build();
        meterManager.record(HistogramEnum.DELIVERY_LATENCY, attributes, latency);
    }

    private void doBeforeConsumeMessage(MessageInterceptorContext context, List<GeneralMessage> messages) {
        if (messages.isEmpty()) {
            // Should never reach here.
            return;
        }
        String consumerGroup = null;
        if (client instanceof PushConsumer) {
            consumerGroup = ((PushConsumer) client).getConsumerGroup();
        }
        if (null == consumerGroup) {
            log.error("[Bug] consumerGroup is not recognized, clientId={}", client.getClientId());
            return;
        }
        final GeneralMessage message = messages.iterator().next();
        final Optional<Long> optionalDecodeTimestamp = message.getDecodeTimestamp();
        if (!optionalDecodeTimestamp.isPresent()) {
            return;
        }
        final long decodeTimestamp = optionalDecodeTimestamp.get();
        Attributes attributes = Attributes.builder().put(MetricLabels.TOPIC, message.getTopic())
            .put(MetricLabels.CONSUMER_GROUP, consumerGroup)
            .put(MetricLabels.CLIENT_ID, client.getClientId().toString()).build();
        final long latency = System.currentTimeMillis() - decodeTimestamp;
        meterManager.record(HistogramEnum.AWAIT_TIME, attributes, latency);
        // Record the time before consuming message.
        context.putAttribute(CONSUME_STOPWATCH_KEY, Attribute.create(Stopwatch.createStarted()));
    }

    private void doAfterConsumeMessage(MessageInterceptorContext context, List<GeneralMessage> messages) {
        if (!(client instanceof PushConsumer)) {
            // Should never reach here.
            log.error("[Bug] current client is not push consumer, clientId={}", client.getClientId());
            return;
        }
        final Attribute<Stopwatch> stopwatchAttr = context.getAttribute(CONSUME_STOPWATCH_KEY);
        if (null == stopwatchAttr) {
            // Should never reach here.
            return;
        }
        PushConsumer pushConsumer = (PushConsumer) client;
        final MessageHookPointsStatus status = context.getStatus();
        for (GeneralMessage message : messages) {
            InvocationStatus invocationStatus = MessageHookPointsStatus.OK.equals(status) ? InvocationStatus.SUCCESS :
                InvocationStatus.FAILURE;
            Attributes attributes = Attributes.builder().put(MetricLabels.TOPIC, message.getTopic())
                .put(MetricLabels.CONSUMER_GROUP, pushConsumer.getConsumerGroup())
                .put(MetricLabels.CLIENT_ID, client.getClientId().toString())
                .put(MetricLabels.INVOCATION_STATUS, invocationStatus.getName())
                .build();
            final Duration duration = stopwatchAttr.get().elapsed();
            meterManager.record(HistogramEnum.PROCESS_TIME, attributes, duration.toMillis());
        }
    }

    @Override
    public void doBefore(MessageInterceptorContext context, List<GeneralMessage> messages) {
        if (!meterManager.isEnabled()) {
            return;
        }
        final MessageHookPoints hookPoints = context.getMessageHookPoints();
        switch (hookPoints) {
            case SEND: {
                doBeforeSendMessage(context);
                break;
            }
            case CONSUME: {
                doBeforeConsumeMessage(context, messages);
                break;
            }
            default:
                break;
        }
    }

    @Override
    public void doAfter(MessageInterceptorContext context, List<GeneralMessage> messages) {
        if (!meterManager.isEnabled()) {
            return;
        }
        final MessageHookPoints hookPoints = context.getMessageHookPoints();
        switch (hookPoints) {
            case SEND: {
                doAfterSendMessage(context, messages);
                break;
            }
            case RECEIVE: {
                doAfterReceiveMessage(messages);
                break;
            }
            case CONSUME: {
                doAfterConsumeMessage(context, messages);
                break;
            }
            default:
                break;
        }
    }
}

File: php/MessageMeterInterceptor.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq;

use Apache\Rocketmq\Message\MessageInterface;
use Apache\Rocketmq\Logger;

/**
 * Message Meter Interceptor - Automatically collects message metrics
 * 
 * This interceptor records:
 * - Send cost time (SEND_BEFORE/SEND_AFTER)
 * - Delivery latency (RECEIVE)
 * - Consume await time (CONSUME_BEFORE)
 * - Consume process time (CONSUME_AFTER)
 * 
 * Reference: Java MessageMeterInterceptor
 */
class MessageMeterInterceptor implements MessageInterceptor {
    /**
     * @var AttributeKey Key for send stopwatch
     */
    private static $SEND_STOPWATCH_KEY;
    
    /**
     * @var AttributeKey Key for consume stopwatch
     */
    private static $CONSUME_STOPWATCH_KEY;
    
    /**
     * @var MetricsCollector Metrics collector instance
     */
    private $metricsCollector;
    
    /**
     * @var string Client ID
     */
    private $clientId;
    
    /**
     * @var string|null Consumer group (for consumers)
     */
    private $consumerGroup;
    
    /**
     * Initialize static properties
     */
    public function __construct(MetricsCollector $metricsCollector, string $clientId, ?string $consumerGroup = null) {
        if (self::$SEND_STOPWATCH_KEY === null) {
            self::$SEND_STOPWATCH_KEY = AttributeKey::create('send_stopwatch');
        }
        if (self::$CONSUME_STOPWATCH_KEY === null) {
            self::$CONSUME_STOPWATCH_KEY = AttributeKey::create('consume_stopwatch');
        }
        
        $this->metricsCollector = $metricsCollector;
        $this->clientId = $clientId;
        $this->consumerGroup = $consumerGroup;
    }
    
    /**
     * {@inheritdoc}
     */
    public function doBefore(MessageInterceptorContextInterface $context, array $messages): void {
        $hookPoints = $context->getMessageHookPoints();
        
        switch ($hookPoints) {
            case MessageHookPoints::SEND_BEFORE:
                $this->doBeforeSendMessage($context);
                break;
                
            case MessageHookPoints::RECEIVE_BEFORE:
                $this->doBeforeReceiveMessage($context);
                break;
                
            case MessageHookPoints::CONSUME_BEFORE:
                $this->doBeforeConsumeMessage($context, $messages);
                break;
                
            case MessageHookPoints::ACK_BEFORE:
                $this->doBeforeAckMessage($context);
                break;
                
            default:
                // Do nothing for other hook points
                break;
        }
    }
    
    /**
     * {@inheritdoc}
     */
    public function doAfter(MessageInterceptorContextInterface $context, array $messages): void {
        $hookPoints = $context->getMessageHookPoints();
        
        switch ($hookPoints) {
            case MessageHookPoints::SEND_AFTER:
                $this->doAfterSendMessage($context, $messages);
                break;
                
            case MessageHookPoints::RECEIVE_AFTER:
                $this->doAfterReceiveMessage($context, $messages);
                break;
                
            case MessageHookPoints::CONSUME_AFTER:
                $this->doAfterConsumeMessage($context, $messages);
                break;
                
            case MessageHookPoints::ACK_AFTER:
                $this->doAfterAckMessage($context, $messages);
                break;
                
            default:
                // Do nothing for other hook points
                break;
        }
    }
    
    /**
     * Record time before sending message
     * 
     * @param MessageInterceptorContextInterface $context Interceptor context
     * @return void
     */
    private function doBeforeSendMessage(MessageInterceptorContextInterface $context): void {
        // Record the start time for measuring send duration
        $startTime = microtime(true);
        $context->putAttribute(
            self::$SEND_STOPWATCH_KEY,
            Attribute::create($startTime)
        );
    }
    
    /**
     * Record send cost time after sending message
     * 
     * @param MessageInterceptorContextInterface $context Interceptor context
     * @param \Apache\Rocketmq\Message\Message[] $messages Sent messages
     * @return void
     */
    private function doAfterSendMessage(MessageInterceptorContextInterface $context, array $messages): void {
        $stopwatchAttr = $context->getAttribute(self::$SEND_STOPWATCH_KEY);
        if ($stopwatchAttr === null) {
            // Should never reach here
            Logger::warn("Send stopwatch attribute not found, clientId={}", [$this->clientId]);
            return;
        }
        
        $startTime = $stopwatchAttr->get();
        $endTime = microtime(true);
        $durationMs = ($endTime - $startTime) * 1000; // Convert to milliseconds
        
        $status = $context->getStatus() === MessageHookPointsStatus::OK
            ? InvocationStatus::SUCCESS
            : InvocationStatus::FAILURE;
        
        foreach ($messages as $message) {
            $labels = [
                MetricLabels::TOPIC => $message->getTopic(),
                MetricLabels::CLIENT_ID => $this->clientId,
                MetricLabels::INVOCATION_STATUS => $status,
            ];
            
            $this->metricsCollector->observeHistogram(
                HistogramEnum::SEND_COST_TIME,
                $labels,
                $durationMs
            );
            
            Logger::debug("Recorded send metric: topic={}, status={}, duration={}ms", [
                $message->getTopic(),
                $status,
                round($durationMs, 2)
            ]);
        }
    }
    
    /**
     * Record consume await time before consuming message
     * 
     * @param MessageInterceptorContextInterface $context Interceptor context
     * @param \Apache\Rocketmq\Message\Message[] $messages Messages to consume
     * @return void
     */
    private function doBeforeConsumeMessage(MessageInterceptorContextInterface $context, array $messages): void {
        if (empty($messages)) {
            return;
        }
        
        if ($this->consumerGroup === null) {
            Logger::error("[Bug] consumerGroup is not recognized, clientId={}", [$this->clientId]);
            return;
        }
        
        $message = $messages[0];
        
        // Try to get decode timestamp from message properties
        $decodeTimestamp = $this->getDecodeTimestamp($message);
        if ($decodeTimestamp !== null) {
            $currentTimeMs = microtime(true) * 1000;
            $awaitTimeMs = $currentTimeMs - $decodeTimestamp;
            
            if ($awaitTimeMs >= 0) {
                $labels = [
                    MetricLabels::TOPIC => $message->getTopic(),
                    MetricLabels::CONSUMER_GROUP => $this->consumerGroup,
                    MetricLabels::CLIENT_ID => $this->clientId,
                ];
                
                $this->metricsCollector->observeHistogram(
                    HistogramEnum::AWAIT_TIME,
                    $labels,
                    $awaitTimeMs
                );
                
                Logger::debug("Recorded await time metric: topic={}, awaitTime={}ms", [
                    $message->getTopic(),
                    round($awaitTimeMs, 2)
                ]);
            } else {
                Logger::debug("Await time is negative, awaitTime={}ms", [$awaitTimeMs]);
            }
        }
        
        // Record the start time for measuring consume duration
        $startTime = microtime(true);
        $context->putAttribute(
            self::$CONSUME_STOPWATCH_KEY,
            Attribute::create($startTime)
        );
    }
    
    /**
     * Record consume process time after consuming message
     * 
     * @param MessageInterceptorContextInterface $context Interceptor context
     * @param \Apache\Rocketmq\Message\Message[] $messages Consumed messages
     * @return void
     */
    private function doAfterConsumeMessage(MessageInterceptorContextInterface $context, array $messages): void {
        if ($this->consumerGroup === null) {
            Logger::error("[Bug] consumerGroup is not recognized, clientId={}", [$this->clientId]);
            return;
        }
        
        $stopwatchAttr = $context->getAttribute(self::$CONSUME_STOPWATCH_KEY);
        if ($stopwatchAttr === null) {
            // Should never reach here
            Logger::warn("Consume stopwatch attribute not found, clientId={}", [$this->clientId]);
            return;
        }
        
        $startTime = $stopwatchAttr->get();
        $endTime = microtime(true);
        $durationMs = ($endTime - $startTime) * 1000; // Convert to milliseconds
        
        $status = $context->getStatus() === MessageHookPointsStatus::OK
            ? InvocationStatus::SUCCESS
            : InvocationStatus::FAILURE;
        
        foreach ($messages as $message) {
            $labels = [
                MetricLabels::TOPIC => $message->getTopic(),
                MetricLabels::CONSUMER_GROUP => $this->consumerGroup,
                MetricLabels::CLIENT_ID => $this->clientId,
                MetricLabels::INVOCATION_STATUS => $status,
            ];
            
            $this->metricsCollector->observeHistogram(
                HistogramEnum::PROCESS_TIME,
                $labels,
                $durationMs
            );
            
            Logger::debug("Recorded process time metric: topic={}, status={}, duration={}ms", [
                $message->getTopic(),
                $status,
                round($durationMs, 2)
            ]);
        }
    }
    
    /**
     * Get decode timestamp from message
     * 
     * @param \Apache\Rocketmq\Message\Message $message Message
     * @return float|null Decode timestamp in milliseconds, or null if not available
     */
    private function getDecodeTimestamp(\Apache\Rocketmq\Message\Message $message): ?float {
        // Try to get from system properties
        try {
            $properties = $message->getProperties();
            if (isset($properties['DECODE_TIMESTAMP'])) {
                return floatval($properties['DECODE_TIMESTAMP']);
            }
            
            // Alternative: check for BORN_TIMESTAMP
            if (isset($properties['BORN_TIMESTAMP'])) {
                return floatval($properties['BORN_TIMESTAMP']);
            }
        } catch (\Exception $e) {
            Logger::debug("Failed to get decode timestamp: {}", [$e->getMessage()]);
        }
        
        return null;
    }
    
    /**
     * Record time before receiving message
     * 
     * @param MessageInterceptorContextInterface $context Interceptor context
     * @return void
     */
    private function doBeforeReceiveMessage(MessageInterceptorContextInterface $context): void {
        // Record the start time for measuring receive duration
        $startTime = microtime(true);
        $context->putAttribute(
            self::$SEND_STOPWATCH_KEY,
            Attribute::create($startTime)
        );
    }
    
    /**
     * Record receive metrics after receiving message
     * 
     * @param MessageInterceptorContextInterface $context Interceptor context
     * @param MessageInterface[] $messages Received messages
     * @return void
     */
    private function doAfterReceiveMessage(MessageInterceptorContextInterface $context, array $messages): void {
        if (empty($messages)) {
            return;
        }
        
        $stopwatchAttr = $context->getAttribute(self::$SEND_STOPWATCH_KEY);
        if ($stopwatchAttr === null) {
            Logger::warn("Receive stopwatch attribute not found, clientId={}", [$this->clientId]);
            return;
        }
        
        $startTime = $stopwatchAttr->get();
        $endTime = microtime(true);
        $durationMs = ($endTime - $startTime) * 1000;
        
        $status = $context->getStatus() === MessageHookPointsStatus::OK
            ? InvocationStatus::SUCCESS
            : InvocationStatus::FAILURE;
        
        foreach ($messages as $message) {
            $labels = [
                MetricLabels::TOPIC => $message->getTopic(),
                MetricLabels::CLIENT_ID => $this->clientId,
                MetricLabels::INVOCATION_STATUS => $status,
            ];
            
            $this->metricsCollector->observeHistogram(
                HistogramEnum::RECEIVE_COST_TIME,
                $labels,
                $durationMs
            );
        }
    }
    
    /**
     * Record time before acknowledging message
     * 
     * @param MessageInterceptorContextInterface $context Interceptor context
     * @return void
     */
    private function doBeforeAckMessage(MessageInterceptorContextInterface $context): void {
        $startTime = microtime(true);
        $context->putAttribute(
            self::$SEND_STOPWATCH_KEY,
            Attribute::create($startTime)
        );
    }
    
    /**
     * Record ack metrics after acknowledging message
     * 
     * @param MessageInterceptorContextInterface $context Interceptor context
     * @param MessageInterface[] $messages Acknowledged messages
     * @return void
     */
    private function doAfterAckMessage(MessageInterceptorContextInterface $context, array $messages): void {
        if (empty($messages)) {
            return;
        }
        
        $stopwatchAttr = $context->getAttribute(self::$SEND_STOPWATCH_KEY);
        if ($stopwatchAttr === null) {
            return;
        }
        
        $startTime = $stopwatchAttr->get();
        $endTime = microtime(true);
        $durationMs = ($endTime - $startTime) * 1000;
        
        $status = $context->getStatus() === MessageHookPointsStatus::OK
            ? InvocationStatus::SUCCESS
            : InvocationStatus::FAILURE;
        
        foreach ($messages as $message) {
            $labels = [
                MetricLabels::TOPIC => $message->getTopic(),
                MetricLabels::CLIENT_ID => $this->clientId,
                MetricLabels::INVOCATION_STATUS => $status,
            ];
            
            $this->metricsCollector->observeHistogram(
                HistogramEnum::ACK_COST_TIME,
                $labels,
                $durationMs
            );
        }
    }
}
